### PR TITLE
fix(highlights): Add TSNamespace highlight

### PR DESCRIPTION
**Pull request by theHamsta**
Start adding highlights for
     - C++
     - Rust (including other scoped_identifier/scoped_type_identifier fixes)
     - JS (only namespace_import)

Addresses #516

Vim has no dedicated hl-group for modules/namespace continued to use Constant (open for better suggestions)

#453 could be useful for Rust highlights. In the queries it can almost always be `scoped_identifier` or `scoped_type_identifier` as a parent.

### CODE REVIEW

**Comment by bluz71** [comment]
> Vim has no dedicated hl-group for modules/namespace continued to use Constant (open for better suggestions)

I am now of the opinion that `Structure` is the appropriate generic highlight group to use as the default for `TSNamespace`.

From the [Namespace](https://en.wikipedia.org/wiki/Namespace) Wikipedia entry:

> Namespaces are commonly **structured** as hierarchies to allow reuse of names in different contexts.

> It is possible to use anonymous **structs** as namespaces in C since C99

Yes, I am stretching it, but we can justify the use of `Structure` for namespaces via the above Wikipedia notes. I feel it is a better group than `Constant` for namespaces.

**Comment by bluz71** [comment]
I just test @theHamsta's new `TSNamespace` branch and it works very well for C++ and Rust.

A definite improvement when theming Treesitter.

Many thanks.

**Comment by theHamsta** [comment]
The documentation for `TSStructure` was not really ideal. Nobody knew what is was for. So it only had one usage.

**Comment by bluz71** [comment]
> The documentation for `TSStructure` was not really ideal. Nobody knew what is was for. So it only had one usage.

`TSStructure` probably shouldn't even exist if it only has one use in OCaml files. It may well be the case the OCaml module names (currently the only use of `TSStructure`) could go over to `TSNamespace`? I don't know OCaml well enough to know if that is a good or terrible idea.

Anyway, `TSNamespace` defaulting to `Structure` seems a reasonable choice.

**Comment by theHamsta** [comment]
Vim doc reads:

```
	 Structure	struct, union, enum, etc.
```

**Comment by bluz71** [comment]
Those Vim docs about syntax groups where likely written 20 years ago when Bram only cared about the C language.

The docs for `Constant` state:

`Constant  any constant`

It's a choice about which default makes the makes sense for `TSNamespace` and will look pretty good with most existing Vim themes. Using the same color for `struct, union, enum` will produce a good result.

**Comment by theHamsta** [comment]
Most colorschemes I experiment with just use keyword highlighting for structure. Not a good choice for an identifier also because it will be the same color as `namespace` itself.

My choice came mainly from using the same highlights as `vim-modern-cpp`
```
hi def link cppSTLnamespace    Constant
```

**Comment by bluz71** [comment]
Ok, that’s a reasonable point, I step back.

Note, with my themes I am already setting custom `TSNamespace` colors in anticipation of this enhancement. Result is nice.

Best regards.

File: plugin/nvim-treesitter.vim [hunk 1]
```diff
@@ -41,6 +41,7 @@ highlight default link TSProperty Identifier
 highlight default link TSConstructor Special
 highlight default link TSAnnotation PreProc
 highlight default link TSAttribute PreProc
+highlight default link TSNamespace Constant
```
Comment (vigoux) [inline plugin/nvim-treesitter.vim:44]: Why adding twice?

Comment (theHamsta) [reply plugin/nvim-treesitter.vim:44]: maybe that was why it was still a draft :-)

File: queries/rust/highlights.scm [hunk 2]
```diff
@@ -60,8 +81,10 @@
   (#eq? @type "derive"))
 
 (macro_invocation
-  macro: (identifier) @function.macro
-  "!" @function.macro)
```
Comment (theHamsta) [inline queries/rust/highlights.scm:64]: "!" was overwritten anyways in the current queries.

File: queries/rust/highlights.scm [hunk 3]
```diff
@@ -36,13 +40,30 @@
...
+(scoped_identifier
+  path: (identifier) @namespace)
+(scoped_identifier
+ (scoped_identifier
+  name: (identifier) @namespace))
+(scoped_type_identifier
+  path: (identifier) @namespace)
+(scoped_type_identifier
+ (scoped_identifier
+  name: (identifier) @namespace))
```
Comment (theHamsta) [inline queries/rust/highlights.scm:52]: It's possible that it was a draft to wait for #453 .
As far as I understood you can do
```lisp
[(scoped_identifier
  path: (identifier) @namespace)
(scoped_type_identifier
  path: (identifier) @namespace)]
```
but not 

```lisp
([scoped_identifier scoped_type_identifier]
  path: (identifier) @namespace)
```
after the emacs tree-sitter creator complained. This is partially addressed by matching on supernodes.

**Review by vigoux — APPROVED** [review]

**Comment by undu** [comment]
Regarding ocaml and the usage of structure: with regex-based highlighting module names are regarded as includes, structures matched that in the colorscheme I used, seeing that it kept the colour and more-or-less matched the namespace use-case I I ended up using in ocaml.

I would ask to use `Include` instead of `Constant`, as namespaces/modules are more related to including than to a constant. At least for ocaml code choosing the same highlight for namespaces and constants would make them stand out less in the code, when they are both important to recognise.

File: plugin/nvim-treesitter.vim (hunk 1 above)
Comment (vigoux) [inline plugin/nvim-treesitter.vim:44]: As @undu suggested, it sounds better to have that highlighted as `Include` instead of `Constant`.

Comment (theHamsta) [reply plugin/nvim-treesitter.vim]: tested the last 24h with different color schemes. `Include` looks generally good although it is close to functions in some color schemes.

**Pull request merged**